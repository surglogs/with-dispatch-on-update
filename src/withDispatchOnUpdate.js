import { isPlainObject } from 'is-plain-object'
import all from 'ramda/src/all'
import isNil from 'ramda/src/isNil'
import map from 'ramda/src/map'
import not from 'ramda/src/not'
import pipe from 'ramda/src/pipe'
import reject from 'ramda/src/reject'
import { connect } from 'react-redux'

const isFunction = (value) => {
  return typeof value === 'function'
}

const hasRequiredProps = (props, keys) => {
  return pipe(
    reject(isFunction),
    map((prop) => props[prop]),
    all(pipe(isNil, not)),
  )(keys)
}

const getArgValues = (props, keys) => {
  return map((prop) => {
    return isFunction(prop) ? prop(props) : props[prop]
  }, keys)
}

const CONNECTED_PROPS = 'CONNECTED_PROPS'

const withDispatchOnUpdate = ({
  action,
  condition = () => true,
  args = [],
  connector = {},
  shouldRequireAllProps = true,
}) => {
  return connect(
    (state, props) => ({
      [CONNECTED_PROPS]: isPlainObject(connector) ? map((f) => f(state, props), connector) : connector,
      state,
    }),
    { boundAction: action },
    ({ [CONNECTED_PROPS]: connectedProps, state }, { boundAction }, props) => {
      const mergedProps = { ...props, ...connectedProps }
      if ((!shouldRequireAllProps || hasRequiredProps(mergedProps, args)) && condition(state, mergedProps)) {
        boundAction(...getArgValues(mergedProps, args))
      }

      return props
    },
  )
}

export default withDispatchOnUpdate
