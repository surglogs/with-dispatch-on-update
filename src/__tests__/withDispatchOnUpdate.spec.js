import React from 'react'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { mount, configure, render } from 'enzyme'
import sinon from 'sinon'
import { compose, withState } from 'recompose'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import withDispatchOnUpdate from '../index'

configure({ adapter: new Adapter() })

const MOCK_ACTION = 'MOCK_ACTION'
const CLEAR_STATE = 'CLEAR_STATE'
const CLEAR_TODOS = 'CLEAR_TODOS'
const SET_CAR = 'SET_CAR'

const buildComponentWithStore = (Component, store) => props => {
  return (
    <Provider {...{ store }}>
      <Component {...props} />
    </Provider>
  )
}

describe('withDispatchOnUpdate HOC', () => {
  it('calls action if and only if component updates and condition is met', () => {
    const component = sinon.spy(() => null)

    const mockAction = sinon.spy(() => {
      return {
        type: MOCK_ACTION,
        payload: {},
      }
    })

    const clearStateAction = () => {
      return {
        type: CLEAR_STATE,
        payload: {},
      }
    }

    const initialState = {
      todos: null,
    }

    const reducer = (state, action) => {
      switch (action.type) {
        case MOCK_ACTION:
          return { ...state, todos: [] }
        case CLEAR_STATE:
          return initialState
        default:
          return state
      }
    }

    const store = createStore(reducer, initialState)

    const shouldDispatchAction = (state, props) => {
      const { todos } = state
      const { showTodos } = props

      return !todos && showTodos === true
    }

    const Todos = compose(
      withState('showTodos', 'setShowTodos', false),
      withDispatchOnUpdate({
        action: mockAction,
        condition: shouldDispatchAction,
        connector: {
          todos: state => state.todos,
        },
      }),
    )(component)

    const ComponentWithStore = buildComponentWithStore(Todos, store)
    mount(<ComponentWithStore />)

    const { setShowTodos } = component.firstCall.args[0]
    expect(mockAction.callCount).toBe(0)

    setShowTodos('No, do not show them!')
    expect(mockAction.callCount).toBe(0)

    setShowTodos(true)
    expect(mockAction.callCount).toBe(1)
    expect(mockAction.lastCall.args).toEqual([])

    setShowTodos(false)
    setShowTodos(true)
    expect(mockAction.callCount).toBe(1)

    store.dispatch(clearStateAction())
    expect(mockAction.callCount).toBe(2)
  })

  it('passes action arguments correctly', () => {
    const component = sinon.spy(() => null)

    const mockAction = sinon.spy((car, name, page) => {
      return {
        type: MOCK_ACTION,
        payload: { car, name, page },
      }
    })

    const clearTodosAction = () => {
      return {
        type: CLEAR_TODOS,
        payload: {},
      }
    }

    const setCarAction = car => {
      return {
        type: SET_CAR,
        payload: { car },
      }
    }

    const initialState = {
      todos: null,
      car: undefined,
      name: 'Gatsby',
    }

    const reducer = (state, action) => {
      switch (action.type) {
        case MOCK_ACTION:
          return { ...state, todos: [] }
        case CLEAR_TODOS:
          return { ...state, todos: null }
        case SET_CAR:
          return { ...state, car: action.payload.car }
        default:
          return state
      }
    }

    const store = createStore(reducer, initialState)

    const shouldDispatchAction = (state, props) => {
      const { todos } = state
      const { showTodos } = props

      return !todos && showTodos === true
    }

    const Todos = compose(
      withState('showTodos', 'setShowTodos', false),
      withState('page', 'setPage', null),
      withDispatchOnUpdate({
        action: mockAction,
        condition: shouldDispatchAction,
        args: [({ car }) => car, 'name', 'page'],
        connector: {
          todos: state => state.todos,
          car: state => state.car,
          name: (state, props) => props.prefix + ' ' + state.name,
        },
      }),
    )(component)

    const ComponentWithStore = buildComponentWithStore(Todos, store)
    mount(<ComponentWithStore {...{ prefix: 'The Great' }} />)

    const { setShowTodos, setPage } = component.firstCall.args[0]
    expect(mockAction.callCount).toBe(0)

    setShowTodos(true)
    expect(mockAction.callCount).toBe(0)

    setPage(1)
    expect(mockAction.callCount).toBe(1)
    expect(mockAction.lastCall.args).toEqual([undefined, 'The Great Gatsby', 1])

    store.dispatch(setCarAction('Volkswagen Beetle'))
    expect(mockAction.callCount).toBe(1)

    store.dispatch(clearTodosAction())
    expect(mockAction.callCount).toBe(2)
    expect(mockAction.lastCall.args).toEqual([
      'Volkswagen Beetle',
      'The Great Gatsby',
      1,
    ])
  })

  it('renders children correctly', () => {
    const component = ({ children }) => {
      return (
        <div>
          I am a parent!
          {children}
        </div>
      )
    }

    const mockAction = sinon.spy(() => {
      return {
        type: MOCK_ACTION,
        payload: {},
      }
    })

    const initialState = {
      todos: null,
    }

    const reducer = (state, action) => {
      switch (action.type) {
        case MOCK_ACTION:
          return { ...state, todos: [] }
        case CLEAR_STATE:
          return initialState
        default:
          return state
      }
    }

    const store = createStore(reducer, initialState)

    const shouldDispatchAction = (state, props) => {
      const { todos } = state
      const { showTodos } = props

      return !todos && showTodos === true
    }

    const Todos = compose(
      withState('showTodos', 'setShowTodos', false),
      withDispatchOnUpdate({
        action: mockAction,
        condition: shouldDispatchAction,
        connector: {
          todos: state => state.todos,
        },
      }),
    )(component)

    const TodosWithChildren = props => {
      return (
        <Todos>
          <span>I am a child!</span>
        </Todos>
      )
    }

    const ComponentWithStore = buildComponentWithStore(TodosWithChildren, store)
    const renderedComponent = render(<ComponentWithStore />)

    expect(toJson(renderedComponent)).toMatchSnapshot()
  })
})
