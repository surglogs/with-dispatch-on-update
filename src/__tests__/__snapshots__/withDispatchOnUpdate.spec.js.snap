// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDispatchOnUpdate HOC renders children correctly 1`] = `
<div>
  I am a parent!
  <span>
    I am a child!
  </span>
</div>
`;
